src/libsystemd-network/sd-dhcp-client.c

```diff
@@ -1465,10 +1507,28 @@ static int client_handle_offer(sd_dhcp_client *client, DHCPMessage *offer, size_
 
         r = dhcp_option_parse(offer, len, dhcp_lease_parse_options, lease, NULL);
         if (r != DHCP_OFFER) {
-                log_dhcp_client(client, "received message was not an OFFER, ignoring");
-                return -ENOMSG;
+                if (r == -ENOMSG && client->bootp) {
+                        // Treat a non-DHCP BOOTREPLY like a DHCP ACK, so keep processing
+                        log_dhcp_client(client, "BOOTREPLY received");
+                        r = DHCP_ACK;
+                } else {
+                        log_dhcp_client(client, "received message was not an OFFER, ignoring");
+                        return -ENOMSG;
+                }
         }
 
+        if (client->bootp) {
+                // This was USEC_INFINITY in systemd v258, but that's a long unsigned int; casting doesn't seem like a good idea
+                lease->lifetime = UINT32_MAX;
+                log_dhcp_client(client, "BOOTP identified, using infinite lease. BOOTP siaddr=(%#x), DHCP Server Identifier=(%#x)",
+                        offer->siaddr,
+                        lease->server_address);
+
+                lease->server_address = offer->siaddr || lease->server_address;
```

Comment:
```suggestion
                lease->server_address = offer->siaddr ? offer->siaddr : lease->server_address;
```
`||` in C returns `0` or `1`, not any of its operands